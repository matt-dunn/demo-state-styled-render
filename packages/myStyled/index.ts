/** !
 * Copyright (c) 2020, Matt Dunn
 *
 * @author Matt Dunn (https://matt-dunn.github.io/)
 * @licence MIT
 */

export { myStyled as default } from "./myStyled";

export { css, cssWithProps } from "./css";
