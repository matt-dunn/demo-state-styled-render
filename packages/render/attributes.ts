/** !
 * Copyright (c) 2020, Matt Dunn
 *
 * @author Matt Dunn (https://matt-dunn.github.io/)
 * @licence MIT
 */

import { AnyFunc, HTMLElementMap, Props } from "./types";
import { propertyMap } from "./utils";

export const setAttribute = (
  element: HTMLElementMap,
  name: string,
  value: string | AnyFunc
) => {
  const propertyName = propertyMap(name);

  if (typeof value === "function") {
    element[propertyName] = value;
  } else {
    if (value) {
      element.setAttribute(propertyName, value);
    } else {
      element.removeAttribute(propertyName);
    }
    element[propertyName] = value;
  }
};

export const setAttributes = (element: HTMLElement, props: Props) =>
  Object.keys(props || {}).forEach((name) =>
    setAttribute(element, name, props[name])
  );

export const removeAttribute = (element: HTMLElement, name: string) =>
  element.removeAttribute(propertyMap(name));

export const updateAttribute = (
  element: HTMLElement,
  name: string,
  value: string,
  prevValue: string
) => {
  if (value !== prevValue) {
    if (value === undefined) {
      removeAttribute(element, name);
    } else if (!prevValue || value !== prevValue) {
      setAttribute(element, name, value);
    }
  }
};

export const updateAttributes = (
  element: HTMLElement,
  props: Props,
  prevProps: Props = {}
) =>
  Object.keys({ ...props, ...prevProps }).forEach((name) =>
    updateAttribute(element, name, props[name], prevProps[name])
  );
