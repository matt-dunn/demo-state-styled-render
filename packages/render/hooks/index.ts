/** !
 * Copyright (c) 2020, Matt Dunn
 *
 * @author Matt Dunn (https://matt-dunn.github.io/)
 * @licence MIT
 */

import { AnyCallback } from "../types";

export type HookID = number;

type HookOptions = {
  render: () => void;
};

export type HookImplementation = {
  type: string;
  cleanup?: AnyCallback;
};

export type HookImplementations = Array<HookImplementation | null>;

export class State<T> extends Array<HookContainer<T>> {
  constructor(...items: HookContainer<T>[]) {
    super(...items);

    Object.setPrototypeOf(this, State.prototype);
  }

  byType<T extends HookImplementation>(
    ...types: Omit<HookImplementation, "id">[]
  ): T[] {
    const filterType = types.map(({ type }) => type);
    return this.filter(
      (item) =>
        (item._value && filterType.indexOf(item._value.type) !== -1) || false
    ).map((item) => item._value) as T[];
  }
}

type Hook<T> = {
  index: number;
  state: State<T>;
  options: HookOptions;
};

export type Hooks<T> = Hook<T>[] & {
  [key: number]: Hook<T>;
};

export type HookContainer<T> = {
  _value: HookImplementation | null;
  options: HookOptions;
  getValue: () => T | undefined;
  setValue: (value: Omit<T, "id">) => T;
};

type HooksContainer = {
  setInsertMode: (insert: boolean) => void;
  register: (options: HookOptions) => HookID;
  setActive: (id: HookID) => void;
  getActive: <T>() => Hook<T>;
  getCurrent: <T = any>() => HookContainer<T>;
  inspectActive: <T = any>() => HookContainer<T> | undefined;
  beginCollect: () => void;
  collect: () => State<any>;
  removeHooks: (count: number) => void;
};

const hook = <T>(options: HookOptions): Hook<T> => ({
  index: 0,
  state: new State(),
  options,
});

const hookContainer = (hook: Hook<any>): HookContainer<any> => ({
  options: hook.options,
  _value: null,
  getValue() {
    return this._value;
  },
  setValue(value: Omit<HookImplementation, "id">) {
    this._value = value;

    return value;
  },
});

const ActiveHooks = (): HooksContainer => {
  let activeId: HookID | undefined;
  let hooks: Hooks<any> = [];
  let collectedHookStartIndex: number;
  let insertMode = false;

  const getActiveHook = (activeId?: HookID): Hook<any> => {
    if (activeId === undefined) {
      throw new TypeError("No active hooks");
    }

    return hooks[activeId];
  };

  const getActiveHookContainer = (
    activeId?: HookID,
    activeIndex?: number
  ): HookContainer<any> => {
    const hook = getActiveHook(activeId);
    const currentIndex = activeIndex ?? hook.index;

    if (insertMode) {
      hook.state.splice(currentIndex, 0, hookContainer(hook));
    }

    const hookState = hook.state[currentIndex];

    if (!hookState) {
      hook.state[currentIndex] = hookContainer(hook);
    }

    return hookState;
  };

  return {
    setInsertMode(insert) {
      insertMode = insert;
    },
    register(options) {
      hooks = [...hooks, hook(options)];
      return hooks.length - 1;
    },
    setActive: (id) => {
      activeId = id;
      hooks[activeId].index = 0;
    },
    getActive() {
      return getActiveHook(activeId);
    },
    getCurrent() {
      const current = getActiveHookContainer(activeId);
      this.getActive().index++;
      return current;
    },
    inspectActive() {
      return getActiveHookContainer(activeId, this.getActive().index - 1);
    },
    beginCollect() {
      collectedHookStartIndex = this.getActive().index;
    },
    collect() {
      const hook = this.getActive();
      const currentIndex = hook.index;

      if (currentIndex !== collectedHookStartIndex) {
        return new State(
          ...hook.state
            .slice(collectedHookStartIndex, currentIndex)
            .map((state) => state && { ...state })
        );
      }

      return new State();
    },
    removeHooks(count) {
      const hook = this.getActive();
      hook.state.splice(hook.index, count);
    },
  };
};

export default ActiveHooks();

export * from "./useState";
export * from "./useEffect";
export * from "./useContext";
export * from "./useError";
