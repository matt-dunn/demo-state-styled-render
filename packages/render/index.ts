/** !
 * Copyright (c) 2020, Matt Dunn
 *
 * @author Matt Dunn (https://matt-dunn.github.io/)
 * @licence MIT
 */

export { Mx as default } from "./render";

export { Node, Children, FC, HTMLElementMap } from "./types";

export { classnames } from "./utils";

export { jsx, jsxFrag, createElement } from "./element";

export * from "./hooks";
