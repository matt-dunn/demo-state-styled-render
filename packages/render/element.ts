/** !
 * Copyright (c) 2020, Matt Dunn
 *
 * @author Matt Dunn (https://matt-dunn.github.io/)
 * @licence MIT
 */

import {
  AnyNode,
  Children,
  Key,
  Node,
  NODE_TYPE_FRAGMENT,
  NodeType,
  Props,
} from "./types";
import { isNode } from "./utils";
import { setAttributes, updateAttributes } from "./attributes";

export const jsx = (
  type: NodeType,
  props: Props = {},
  ...children: Children
): Node => {
  const { key, ...rest } = props || {};

  return typeof type === "function"
    ? type({
        ...rest,
        children: children?.length === 1 ? children[0] : children,
        key,
      })
    : { type, props: rest, children, key };
};

export const jsxFrag = ({
  children,
  ...props
}: {
  children: Children;
}): Node & { key: Key } => ({
  type: NODE_TYPE_FRAGMENT,
  props: props,
  children: children,
  key: null,
});

(global as any).jsx = jsx;
(global as any).jsxFrag = jsxFrag;

export const createElement = (
  type: NodeType,
  props: Props = {},
  children?: Children | Node
): Node => {
  if (Array.isArray(children)) {
    return jsx(type, props, ...children);
  } else if (children) {
    return jsx(type, props, children);
  } else {
    return jsx(type, props);
  }
};

const createDocumentElement = (node: AnyNode): HTMLElement | Text => {
  if (isNode(node) && typeof node.type === "string") {
    const element = document.createElement(node.type);

    setAttributes(element, node?.props);

    flattenChildren(node.children)
      .map(createDocumentElement)
      .forEach(element.appendChild.bind(element));

    return element;
  } else if (typeof node === "string" || typeof node === "number") {
    return document.createTextNode(node.toString());
  }

  throw new TypeError(
    `Unable to render invalid node ${JSON.stringify(node).substr(0, 20)}`
  );
};

const hasChanged = (node: AnyNode, prevNode: AnyNode) =>
  typeof node !== typeof prevNode ||
  (typeof node === "string" && node !== prevNode) ||
  (isNode(node) && node.type) !== (isNode(prevNode) && prevNode.type);

const updateChildren = (
  element: HTMLElement,
  children: Children,
  prevChildren: Children,
  index: number
) => {
  const nodeChildrenLength = children.length;
  const prevNodeChildrenLength = prevChildren.length;
  for (
    let childIndex = 0;
    childIndex < nodeChildrenLength || childIndex < prevNodeChildrenLength;
    childIndex++
  ) {
    updateTree(
      element.childNodes[index] as HTMLElement,
      children[childIndex],
      prevChildren[childIndex],
      childIndex
    );
  }
};

const flattenChildren = (children: Children = []): Children =>
  children.reduce((children, child) => {
    if (child.type === NODE_TYPE_FRAGMENT) {
      return [...children, ...flattenChildren(child.children)];
    }
    return [...children, child];
  }, [] as Children);

export const updateTree = (
  element: HTMLElement,
  node: Node,
  prevNode?: Node,
  index = 0
) => {
  if (prevNode === undefined) {
    element.appendChild(createDocumentElement(node));
  } else if (node === undefined) {
    element.removeChild(element.childNodes[index]);
  } else if (hasChanged(node, prevNode)) {
    element.replaceChild(
      createDocumentElement(node),
      element.childNodes[index]
    );
  } else if (isNode(node)) {
    updateAttributes(
      element.childNodes[index] as HTMLElement,
      node.props,
      prevNode.props
    );

    updateChildren(
      element,
      flattenChildren(node.children),
      flattenChildren(prevNode.children),
      index
    );
  } else if (element.childNodes[index].nodeValue !== node) {
    element.childNodes[index].nodeValue = node;
  }
};
