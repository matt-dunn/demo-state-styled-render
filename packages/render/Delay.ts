/** !
 * Copyright (c) 2020, Matt Dunn
 *
 * @author Matt Dunn (https://matt-dunn.github.io/)
 * @licence MIT
 */

import {AnyNode, Node} from "./";
import { useEffect, useState } from "./hooks";

type DelayProps = {
  delay?: number;
  children: Node;
};

export const Delay = ({ delay = 500, children }: DelayProps): AnyNode => {
  const [waiting, setWaiting] = useState(delay !== 0);

  useEffect(() => {
    if (waiting) {
      const t = setTimeout(() => {
        setWaiting(false);
      }, delay);

      return () => {
        clearTimeout(t);
      };
    }
  }, [delay]);

  return waiting ? null : children;
};
