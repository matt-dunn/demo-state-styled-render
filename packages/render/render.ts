/** !
 * Copyright (c) 2020, Matt Dunn
 *
 * @author Matt Dunn (https://matt-dunn.github.io/)
 * @licence MIT
 */

import {Actions, Store} from "packages/state";

import {FC, Node} from "./types";
import {updateTree} from "./element";

type UseState = <S = any>(initialState: S) => [S, (newValue: S) => S];

type MxContainer = {
  index: number;
  state: any[];
  currentState: any;
  currentTree?: Node;
  renderTree: () => void;
  useState: UseState;
  render: <S>(store: Store<S>, actions: Actions) => (component: FC) => (mountPoint: HTMLElement) => void;
}

export const Mx: MxContainer = {
  index: 0,
  state: [],
  currentState: undefined,
  currentTree: undefined,
  renderTree: undefined as unknown as () => void,
  useState: <S = any>(initialState: S) => {
    const activeIndex = Mx.index;

    if (Mx.state[activeIndex] === undefined) {
      Mx.state[activeIndex] = initialState;
    }

    const currentState = Mx.state[activeIndex];

    const setter = (newValue: S) => {
      if (Mx.state[activeIndex] !== newValue) {
        Mx.state[activeIndex] = newValue;
        Mx.renderTree();
      }
      return newValue;
    };

    Mx.index++;

    return [currentState, setter];
  },
  render: (store, actions) => component => mountPoint => {
    const dispatchActions = Object.entries(actions).reduce((actions, [key, action]) => ({
      ...actions,
      [key]: (...args: any[]) => store.dispatch(action(...args))
    }), {});

    Mx.renderTree = () => {
      const currentTree = component({...Mx.currentState, ...dispatchActions});
      updateTree(mountPoint, currentTree, Mx.currentTree);

      Mx.currentTree = currentTree;
      Mx.index = 0;
    };

    Mx.currentState = store.getState();

    store.subscribe((state) => {
      Mx.currentState = state;
      Mx.renderTree();
    });

    return Mx.renderTree();
  }
};

export const useState = Mx.useState;
