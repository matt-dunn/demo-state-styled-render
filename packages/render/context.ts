/** !
 * Copyright (c) 2020, Matt Dunn
 *
 * @author Matt Dunn (https://matt-dunn.github.io/)
 * @licence MIT
 */

import { Node } from "packages/render";

import { looseRef } from "./utils";

export const getValueSymbol = Symbol("getValueSymbol");

export type Context<T = any> = {
  [getValueSymbol]: () => T | undefined;
  Provider: ({ value, children }: { value: T; children: Node }) => Node;
};

export const createContext = <T = any>(defaultValue?: T): Context<T> => {
  const currentValue = looseRef(defaultValue);

  return {
    [getValueSymbol]: () => currentValue.current,
    Provider: ({ value, children }) => {
      if (currentValue.current !== value) {
        currentValue.current = value;
      }

      return children;
    },
  };
};
