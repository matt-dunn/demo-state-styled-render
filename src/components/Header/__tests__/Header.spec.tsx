/** !
 * Copyright (c) 2020, Matt Dunn
 *
 * @author Matt Dunn (https://matt-dunn.github.io/)
 * @licence MIT
 */

/** @jsx jsx **/

import { jsx } from "packages/render";
import { mount } from "packages/render/test-utils";

import { Header } from "../Header";

describe("Header: Example of using mount", () => {
  it("should render component", async () => {
    const wrapper = mount(<Header />);

    expect(wrapper).toMatchSnapshot();
  });
});
