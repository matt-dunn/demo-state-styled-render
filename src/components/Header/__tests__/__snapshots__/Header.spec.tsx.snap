// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header: Example of using mount should render component 1`] = `
<header
  class=" ms__blmrZ"
>
  <h1>
    Pure TypeScript implementation of styled JSX functional components and state management
  </h1>
  <aside
    class=" ms__lhcqzR"
  >
    <a
      href="https://matt-dunn.github.io/"
      rel="noopener noreferrer"
      target="_blank"
    >
      <h2>
        Matt Dunn
      </h2>
      <p>
        Senior User Interface / Full Stack Developer
      </p>
    </a>
  </aside>
</header>
`;
